File: backend/src/main/java/com/projeto/agroecologia/domain/service/MonolitoService.java
package com.projeto.agroecologia.domain.service;

import com.projeto.agroecologia.domain.model.Especie;
import com.projeto.agroecologia.domain.model.Monolito;
import com.projeto.agroecologia.domain.model.Parcela;
import com.projeto.agroecologia.domain.model.Tombo;
import com.projeto.agroecologia.domain.repository.EspecieRepository;
import com.projeto.agroecologia.domain.repository.MonolitoRepository;
import com.projeto.agroecologia.domain.repository.ParcelaRepository;
import com.projeto.agroecologia.domain.repository.TomboRepository;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@Service
public class MonolitoService {

    private final MonolitoRepository monolitoRepository;
    private final EspecieRepository especieRepository;
    private final TomboRepository tomboRepository;
    private final ParcelaRepository parcelaRepository;

    public MonolitoService(
            MonolitoRepository monolitoRepository,
            EspecieRepository especieRepository,
            TomboRepository tomboRepository,
            ParcelaRepository parcelaRepository) {
        this.monolitoRepository = monolitoRepository;
        this.especieRepository = especieRepository;
        this.tomboRepository = tomboRepository;
        this.parcelaRepository = parcelaRepository;
    }

    /*
     * ============================
     * MONÓLITO
     * ============================
     */

    public List<Monolito> listarTodos() {
        return monolitoRepository.findAll();
    }

    public Monolito salvar(Monolito monolito) {

        if (monolitoRepository.existsByStationFieldNumber(
                monolito.getStationFieldNumber())) {

            throw new ResponseStatusException(
                    HttpStatus.CONFLICT,
                    "Station Field Number já cadastrado");
        }

        return monolitoRepository.save(monolito);
    }

    public Monolito buscarPorId(Long id) {
        return monolitoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Monólito não encontrado"));
    }

    public void deletar(Long id) {
        if (!monolitoRepository.existsById(id)) {
            throw new RuntimeException("Monólito não encontrado");
        }
        monolitoRepository.deleteById(id);
    }

    @Transactional
    public Monolito vincularParcela(Long monolitoId, Long parcelaId) {
        Monolito monolito = buscarPorId(monolitoId);
        Parcela parcela = parcelaRepository.findById(parcelaId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Parcela não encontrada"));
        monolito.setParcela(parcela);
        return monolitoRepository.save(monolito);
    }

    @Transactional
    public Monolito desvincularParcela(Long monolitoId) {
        Monolito monolito = buscarPorId(monolitoId);
        monolito.setParcela(null);
        return monolitoRepository.save(monolito);
    }

    @Transactional
    public Monolito atualizar(Long id, Monolito dados) {
        Monolito monolito = buscarPorId(id);

        monolito.setStationFieldNumber(dados.getStationFieldNumber());
        monolito.setSamplingNumber(dados.getSamplingNumber());
        monolito.setMetodo(dados.getMetodo());
        monolito.setProfundidadeSolo(dados.getProfundidadeSolo());
        monolito.setDia(dados.getDia());
        monolito.setMes(dados.getMes());
        monolito.setAno(dados.getAno());
        monolito.setCollector(dados.getCollector());
        monolito.setRemarks(dados.getRemarks());
        monolito.setDescription(dados.getDescription());

        if (dados.getParcela() != null) {
            monolito.setParcela(dados.getParcela());
        }

        return monolitoRepository.save(monolito);
    }

    public List<Monolito> buscarPorCollector(String collector) {
        return monolitoRepository.findByCollectorContainingIgnoreCase(collector);
    }

    public List<Monolito> buscarPorMetodo(String metodo) {
        return monolitoRepository.findByMetodoContainingIgnoreCase(metodo);
    }

    public List<String> listarMetodos() {
        return monolitoRepository.listarMetodosDistinct();
    }

    public List<Monolito> buscarMonolitosPorEspecie(String especie) {
        return tomboRepository.buscarMonolitosPorEspecie(especie);
    }

    /*
     * ============================
     * TOMBO (Monólito ↔ Espécie)
     * ============================
     */

    @Transactional
    public void adicionarEspecieComDados(
            Long monolitoId,
            Long especieId,
            Integer abundancia,
            String identificador) {
        Monolito monolito = buscarPorId(monolitoId);

        Especie especie = especieRepository.findById(especieId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Espécie não encontrada"));

        if (tomboRepository.existsByIdentificador(identificador)) {
            throw new ResponseStatusException(
                    HttpStatus.CONFLICT,
                    "Identificador já cadastrado");
        }

        Tombo tombo = new Tombo();
        tombo.setMonolito(monolito);
        tombo.setEspecie(especie);
        tombo.setAbundancia(abundancia);
        tombo.setIdentificador(identificador);

        tomboRepository.save(tombo);
    }

    public boolean existePorStationFieldNumber(String stationFieldNumber) {
        return monolitoRepository.existsByStationFieldNumber(stationFieldNumber);
    }

    public List<Tombo> listarTombos(Long monolitoId) {
        Monolito monolito = buscarPorId(monolitoId);
        return monolito.getTombos();
    }

    @Transactional
    public void removerTombo(Long tomboId) {
        if (!tomboRepository.existsById(tomboId)) {
            throw new RuntimeException("Registro de tombo não encontrado");
        }
        tomboRepository.deleteById(tomboId);
    }

    public List<Tombo> buscarTombosPorIdentificador(String identificador) {
        return tomboRepository.findByIdentificadorContainingIgnoreCase(identificador);
    }

    public boolean existeIdentificador(String identificador) {
        return tomboRepository.existsByIdentificador(identificador);
    }
}

File: frontend/src/app/services/monolito.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Monolito } from '../models/monolito.model';
import { Tombo } from '../models/tombo.model';

@Injectable({
  providedIn: 'root'
})
export class MonolitoService {

  private apiUrl = 'http://localhost:8080/monolitos';

  constructor(private http: HttpClient) {}

  salvar(monolito: Monolito): Observable<Monolito> {
    return this.http.post<Monolito>(this.apiUrl, monolito);
  }

  listar(): Observable<Monolito[]> {
    return this.http.get<Monolito[]>(this.apiUrl);
  }

  listarMetodos(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/metodos`);
  }

  buscarPorMetodo(metodo: string): Observable<Monolito[]> {
    const params = new HttpParams().set('metodo', metodo);
    return this.http.get<Monolito[]>(`${this.apiUrl}/metodo`, { params });
  }

  buscarPorEspecie(especie: string): Observable<Monolito[]> {
    const params = new HttpParams().set('especie', especie);
    return this.http.get<Monolito[]>(`${this.apiUrl}/especie`, { params });
  }

  verificarIdentificadorExistente(identificador: string): Observable<boolean> {
    return this.http.get<boolean>(
      `${this.apiUrl}/tombos/verificar-identificador`,
      { params: { identificador } }
    );
  }

  verificarStationFieldNumber(stationFieldNumber: string): Observable<boolean> {
    return this.http.get<boolean>(
      `${this.apiUrl}/verificar`,
      { params: { stationFieldNumber } }
    );
  }

  buscarPorCollector(collector: string): Observable<Monolito[]> {
    const params = new HttpParams().set('collector', collector);
    return this.http.get<Monolito[]>(
      `${this.apiUrl}/collector`,
      { params }
    );
  }

  buscarTombosPorIdentificador(identificador: string): Observable<Tombo[]> {
    const params = new HttpParams().set('identificador', identificador);
    return this.http.get<Tombo[]>(
      `${this.apiUrl}/tombos/buscar`,
      { params }
    );
  }

  listarTombos(monolitoId: number): Observable<Tombo[]> {
    return this.http.get<Tombo[]>(
      `${this.apiUrl}/${monolitoId}/tombos`
    );
  }

  deletar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/deletar/${id}`);
  }

  adicionarEspecieComDados(
    monolitoId: number,
    especieId: number,
    abundancia: number,
    identificador: string
  ): Observable<void> {
    const url = `${this.apiUrl}/${monolitoId}/especies/${especieId}`;
    return this.http.post<void>(url, { abundancia, identificador });
  }

  removerTombo(tomboId: number): Observable<void> {
    return this.http.delete<void>(
      `${this.apiUrl}/tombos/${tomboId}`
    );
  }
}